File: src/main/java/com/worldscape/voronoi/VoronoiVertex.java
package com.worldscape.voronoi;

import com.worldscape.voronoi.VoronoiEdge;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class VoronoiVertex {
    private static int idCounter = 0;
    private final String id;
    private final double x;
    private final double y;
    private final List<VoronoiEdge> adjacentEdges = new ArrayList<VoronoiEdge>();

    public VoronoiVertex(double x, double y) {
        this.id = "v_" + idCounter++;
        this.x = x;
        this.y = y;
    }

    public VoronoiVertex(String id, double x, double y) {
        this.id = id;
        this.x = x;
        this.y = y;
    }

    public String getId() {
        return this.id;
    }

    public double getX() {
        return this.x;
    }

    public double getY() {
        return this.y;
    }

    public void addAdjacentEdge(VoronoiEdge edge) {
        if (!this.adjacentEdges.contains(edge)) {
            this.adjacentEdges.add(edge);
        }
    }

    public List<VoronoiEdge> getAdjacentEdges() {
        return new ArrayList<VoronoiEdge>(this.adjacentEdges);
    }

    public double distanceTo(VoronoiVertex other) {
        double dx = this.x - other.x;
        double dy = this.y - other.y;
        return Math.sqrt(dx * dx + dy * dy);
    }

    public double distanceTo(double ox, double oy) {
        double dx = this.x - ox;
        double dy = this.y - oy;
        return Math.sqrt(dx * dx + dy * dy);
    }

    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof VoronoiVertex)) {
            return false;
        }
        VoronoiVertex vertex = (VoronoiVertex)o;
        return Double.compare(vertex.x, this.x) == 0 && Double.compare(vertex.y, this.y) == 0;
    }

    public int hashCode() {
        return Objects.hash(this.x, this.y);
    }

    public String toString() {
        return "VoronoiVertex{id='" + this.id + "', x=" + this.x + ", y=" + this.y + "}";
    }
}

File: src/main/kotlin/com/worldscape/voronoi/VoronoiVertex.kt
package com.worldscape.voronoi

class VoronoiVertex(
    val id: String = "v_" + idCounter++,
    val x: Double,
    val y: Double
) {
    // Secondary constructor for Java interop (Java cannot use Kotlin default params)
    // 供 Java 互操作的辅助构造函数
    constructor(px: Double, py: Double) : this(x = px, y = py)

    companion object {
        @JvmField var idCounter: Int = 0
    }

    private val adjacentEdges: MutableList<VoronoiEdge> = ArrayList()

    fun addAdjacentEdge(edge: VoronoiEdge) {
        if (edge !in adjacentEdges) adjacentEdges.add(edge)
    }

    fun getAdjacentEdges(): List<VoronoiEdge> = ArrayList(adjacentEdges)

    fun distanceTo(other: VoronoiVertex): Double = distanceTo(other.x, other.y)

    fun distanceTo(ox: Double, oy: Double): Double {
        val dx = x - ox
        val dy = y - oy
        return Math.sqrt(dx * dx + dy * dy)
    }

    override fun equals(other: Any?): Boolean =
        this === other || (other is VoronoiVertex && x == other.x && y == other.y)

    override fun hashCode(): Int = java.util.Objects.hash(x, y)

    override fun toString(): String = "VoronoiVertex{id='$id', x=$x, y=$y}"
}